### ISSUE TITLE
Issue initializing page object elements with appium and a ProvidedDriver

### ISSUE THREAD

**Issue by dbirch-cognitoiq** [serenity-bdd/serenity-core#1088]
I'm seeing some issues currently using serenity 1.8.20 and appium 6.0.0-BETA1, I was previously using the webdriver.driver=appium property in my serenity.properties file and letting serenity handle my driver instance, this resulted in #993 which is currently being fixed.

As a separate piece of work, I have changed my driver over to use the provided property, with my own class extending DriverSource managing the AppiumDriver instance creation. This now seems to bypass my previous issue, but is still resulting in some issues with page object initialization.

The new issues result in a stack trace similar to:
```
23/01 16:44:13.672 WARN  [main] n.t.c.p.Pages: Failed to instantiate page of type class com.cognitoiq.test.hermesworkflow.pages.subdepot.SubDepotToVanPage ({})
java.lang.IllegalArgumentException: Can not set net.serenitybdd.core.pages.WebElementFacade field com.mypackage.pages.subdepot.SubDepotFormPage.scanResultInfoText to org.openqa.selenium.remote.RemoteWebElement$$EnhancerByCGLIB$$d27c0df4
	at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167) ~[na:1.8.0_144]
	at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171) ~[na:1.8.0_144]
	at sun.reflect.UnsafeObjectFieldAccessorImpl.set(UnsafeObjectFieldAccessorImpl.java:81) ~[na:1.8.0_144]
	at java.lang.reflect.Field.set(Field.java:764) ~[na:1.8.0_144]
	at org.openqa.selenium.support.PageFactory.proxyFields(PageFactory.java:117) ~[selenium-support-3.8.1.jar:na]
	at org.openqa.selenium.support.PageFactory.initElements(PageFactory.java:105) ~[selenium-support-3.8.1.jar:na]
	at net.thucydides.core.annotations.locators.SmartElementProxyCreator.lambda$new$2(SmartElementProxyCreator.java:32) ~[serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.annotations.locators.SmartElementProxyCreator.proxyElements(SmartElementProxyCreator.java:48) ~[serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.webdriver.DefaultPageObjectInitialiser.apply(DefaultPageObjectInitialiser.java:21) ~[serenity-core-1.8.20.jar:1.8.20]
	at net.serenitybdd.core.pages.PageObject.setDriver(PageObject.java:153) ~[serenity-core-1.8.20.jar:1.8.20]
	at net.serenitybdd.core.pages.PageObject.setDriver(PageObject.java:157) ~[serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.pages.Pages.createFromSimpleConstructor(Pages.java:236) [serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.pages.Pages.getCurrentPageOfType(Pages.java:204) [serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.pages.Pages.getPage(Pages.java:90) [serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.steps.PageObjectDependencyInjector.instantiatePageObjectIfNotAssigned(PageObjectDependencyInjector.java:55) [serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.steps.PageObjectDependencyInjector.injectDependenciesInto(PageObjectDependencyInjector.java:33) [serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.steps.StepFactory.injectOtherDependenciesInto(StepFactory.java:175) [serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.steps.StepFactory.instantiateNewStepLibraryFor(StepFactory.java:165) [serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.steps.StepFactory.instantiateNewStepLibraryFor(StepFactory.java:148) [serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.steps.StepFactory.getNewStepLibraryFor(StepFactory.java:95) [serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.steps.IndividualInstancesByDefaultStepCreationStrategy.initiateStepsFor(IndividualInstancesByDefaultStepCreationStrategy.java:16) [serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.steps.StepAnnotations.instantiateAnyUnitiaializedSteps(StepAnnotations.java:75) [serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.steps.StepAnnotations.instanciateScenarioStepFields(StepAnnotations.java:59) [serenity-core-1.8.20.jar:1.8.20]
	at net.thucydides.core.steps.StepAnnotations.injectScenarioStepsInto(StepAnnotations.java:41) [serenity-core-1.8.20.jar:1.8.20]
	at net.serenitybdd.core.Serenity.injectScenarioStepsInto(Serenity.java:135) [serenity-core-1.8.20.jar:1.8.20]
	at net.serenitybdd.core.Serenity.initializeWithNoStepListener(Serenity.java:101) [serenity-core-1.8.20.jar:1.8.20]
	at cucumber.runtime.SerenityObjectFactory.newInstance(SerenityObjectFactory.java:68) [serenity-cucumber-1.6.10.jar:na]
	at cucumber.runtime.SerenityObjectFactory.cacheNewInstance(SerenityObjectFactory.java:51) [serenity-cucumber-1.6.10.jar:na]
	at cucumber.runtime.SerenityObjectFactory.getInstance(SerenityObjectFactory.java:41) [serenity-cucumber-1.6.10.jar:na]
	at cucumber.runtime.java.JavaStepDefinition.execute(JavaStepDefinition.java:37) [cucumber-java-2.3.1.jar:na]
	at cucumber.runtime.StepDefinitionMatch.runStep(StepDefinitionMatch.java:40) [cucumber-core-2.3.0.jar:na]
	at cucumber.api.TestStep.executeStep(TestStep.java:102) [cucumber-core-2.3.0.jar:na]
	at cucumber.api.TestStep.run(TestStep.java:83) [cucumber-core-2.3.0.jar:na]
	at cucumber.api.TestCase.run(TestCase.java:58) [cucumber-core-2.3.0.jar:na]
	at cucumber.runner.Runner.runPickle(Runner.java:80) [cucumber-core-2.3.0.jar:na]
	at cucumber.runtime.junit.PickleRunners$NoStepDescriptions.run(PickleRunners.java:140) [cucumber-junit-2.3.1.jar:na]
	at cucumber.runtime.junit.FeatureRunner.runChild(FeatureRunner.java:68) [cucumber-junit-2.3.1.jar:na]
	at cucumber.runtime.junit.FeatureRunner.runChild(FeatureRunner.java:23) [cucumber-junit-2.3.1.jar:na]
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363) [junit-4.12.jar:4.12]
	at cucumber.runtime.junit.FeatureRunner.run(FeatureRunner.java:73) [cucumber-junit-2.3.1.jar:na]
	at cucumber.api.junit.Cucumber.runChild(Cucumber.java:117) [cucumber-junit-2.3.1.jar:na]
	at cucumber.api.junit.Cucumber.runChild(Cucumber.java:55) [cucumber-junit-2.3.1.jar:na]
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) [junit-4.12.jar:4.12]
	at cucumber.api.junit.Cucumber$1.evaluate(Cucumber.java:126) [cucumber-junit-2.3.1.jar:na]
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363) [junit-4.12.jar:4.12]
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137) [junit-4.12.jar:4.12]
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68) [junit-rt.jar:na]
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47) [junit-rt.jar:na]
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242) [junit-rt.jar:na]
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70) [junit-rt.jar:na]
    When clicking on the "Van to Sub Depot" menu              # NavigationStepDefs.iGoToTheMenu(String)
      net.thucydides.core.pages.WrongPageError: The page object class com.cognitoiq.test.hermesworkflow.pages.subdepot.SubDepotToVanPage looks dodgy:
Failed to instantiate page (java.lang.IllegalArgumentException: Can not set net.serenitybdd.core.pages.WebElementFacade field com.mypackage.pages.subdepot.SubDepotFormPage.scanResultInfoText to org.openqa.selenium.remote.RemoteWebElement$$EnhancerByCGLIB$$d27c0df4)
	at net.thucydides.core.pages.Pages.thisPageObjectLooksDodgy(Pages.java:293)
	at net.thucydides.core.pages.Pages.getCurrentPageOfType(Pages.java:223)
	at net.thucydides.core.pages.Pages.getPage(Pages.java:90)
	at net.thucydides.core.steps.PageObjectDependencyInjector.instantiatePageObjectIfNotAssigned(PageObjectDependencyInjector.java:55)
	at net.thucydides.core.steps.PageObjectDependencyInjector.injectDependenciesInto(PageObjectDependencyInjector.java:33)
	at net.thucydides.core.steps.StepFactory.injectOtherDependenciesInto(StepFactory.java:175)
	at net.thucydides.core.steps.StepFactory.instantiateNewStepLibraryFor(StepFactory.java:165)
	at net.thucydides.core.steps.StepFactory.instantiateNewStepLibraryFor(StepFactory.java:148)
	at net.thucydides.core.steps.StepFactory.getNewStepLibraryFor(StepFactory.java:95)
	at net.thucydides.core.steps.IndividualInstancesByDefaultStepCreationStrategy.initiateStepsFor(IndividualInstancesByDefaultStepCreationStrategy.java:16)
	at net.thucydides.core.steps.StepAnnotations.instantiateAnyUnitiaializedSteps(StepAnnotations.java:75)
	at net.thucydides.core.steps.StepAnnotations.instanciateScenarioStepFields(StepAnnotations.java:59)
	at net.thucydides.core.steps.StepAnnotations.injectScenarioStepsInto(StepAnnotations.java:41)
	at net.serenitybdd.core.Serenity.injectScenarioStepsInto(Serenity.java:135)
	at net.serenitybdd.core.Serenity.initializeWithNoStepListener(Serenity.java:101)
	at cucumber.runtime.SerenityObjectFactory.newInstance(SerenityObjectFactory.java:68)
	at cucumber.runtime.SerenityObjectFactory.cacheNewInstance(SerenityObjectFactory.java:51)
	at cucumber.runtime.SerenityObjectFactory.getInstance(SerenityObjectFactory.java:41)
	at cucumber.runtime.java.JavaStepDefinition.execute(JavaStepDefinition.java:37)
	at cucumber.runtime.StepDefinitionMatch.runStep(StepDefinitionMatch.java:40)
	at cucumber.api.TestStep.executeStep(TestStep.java:102)
	at cucumber.api.TestStep.run(TestStep.java:83)
	at cucumber.api.TestCase.run(TestCase.java:58)
	at cucumber.runner.Runner.runPickle(Runner.java:80)
	at cucumber.runtime.junit.PickleRunners$NoStepDescriptions.run(PickleRunners.java:140)
	at cucumber.runtime.junit.FeatureRunner.runChild(FeatureRunner.java:68)
	at cucumber.runtime.junit.FeatureRunner.runChild(FeatureRunner.java:23)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at cucumber.runtime.junit.FeatureRunner.run(FeatureRunner.java:73)
	at cucumber.api.junit.Cucumber.runChild(Cucumber.java:117)
	at cucumber.api.junit.Cucumber.runChild(Cucumber.java:55)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at cucumber.api.junit.Cucumber$1.evaluate(Cucumber.java:126)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at ✽.clicking on the "Van to Sub Depot" menu(features/subdepot/MainMenu.feature:8)
```

I've debugged through the serenity and appium code, and have found one possible cause for this.

When the AppiumFieldDecorator (called from you SmartElementProxyCreator class)  is running the first line of its constructor which unpacks the web driver from the search context, it is returning the webDriver type as a WebDriverFacade, this class doesn't extend HasSessionDetails, so the platform and automation variables are set to null, these variables are later passed through to the getEnhancedProxy method as part of ElementMap factory method and being instantiated as standard RemoteWebElements types rather than the expected mobile element type.

I from the changes I've seen, updating SmartElementProxyCreator to call getProxiedDriver() rather than just passing through the WebDriverFacade would avoid this issue, but I have no idea what other effects this might have on other ways of executing this code.

Do you have any thoughts on next steps? I could probably put a reproduction scenario together, writing some kind of unit test or even trying to put a PR together for this one.

**Comment by cliviu** [comment]
A new rc version will be available soon

**Comment by dbirch-cognitoiq** [comment]
@cliviu Is it worth me grabbing the current codebase and trying this against head to confirm I no longer see the issue? Been thinking of taking a fork and have a go at fixing some of the issues I'm finding so may as well get used to building serenity :)

**Comment by cliviu** [comment]
my PR is merged (https://github.com/serenity-bdd/serenity-core/commit/4553a4eab1c7ff51a4d86c19c87f847195534245) . I've tested yesterday with my hybrid application and I was able to use WebElementFacade instead of WebElement. Either you wait until the new rc version is published or you build serenity head.

**Comment by cliviu** [comment]
Version 1.8.22-rc.2 is on Bintray

**Comment by cliviu** [comment]
I have also updated the poc project to use the WebElementFacade

**Comment by wakaleo** [comment]
Fixed in 1.9.0

**Comment by PratapSingh09** [comment]
java.lang.IllegalArgumentException: Can not set net.serenitybdd.core.pages.WebElementFacade field

Below Code used:
 @AndroidFindBy(id = "com.awe.wed:id/see_more")
  private WebElementFacade seeMoreIcon;

public class MobilePageObject extends PageObject {

  public MobilePageObject(final WebDriver driver) {

    super(driver, new Predicate<PageObject>() {

      public boolean apply(PageObject page) {

        PageFactory.initElements(
            new AppiumFieldDecorator(((WebDriverFacade) page.getDriver()).getProxiedDriver(),
                page.getImplicitWaitTimeout().in(TimeUnit.SECONDS), TimeUnit.SECONDS),
            page);

        return true;
      }

    });

  }

Error:

[main] WARN net.thucydides.core.pages.Pages - Failed to instantiate page of type class com.itv.android.page.HomePage ({})
java.lang.IllegalArgumentException: Can not set net.serenitybdd.core.pages.WebElementFacade field com.itv.android.page.HomePage.seeMoreIcon to io.appium.java_client.android.AndroidElement$$EnhancerByCGLIB$$3c371404
	at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
	at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArg


Dependency used:
serenity.version>1.2.5-rc.1</serenity.version>
		<serenity.cucumber.version>1.1.28</serenity.cucumber.version>
		<serenity.maven.version>1.2.5-rc.1</serenity.maven.version>

And tryed to changes it to 1.9.0.rc.3 but did not found dependecy

Pom Used:

<properties>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
		<project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
		<java.version>1.8</java.version>
		<serenity.version>1.2.5-rc.1</serenity.version>
		<serenity.cucumber.version>1.1.28</serenity.cucumber.version>
		<serenity.maven.version>1.2.5-rc.1</serenity.maven.version>
		<surefire.rerunFailingTestsCount>0</surefire.rerunFailingTestsCount>
		<maven.test.failure.ignore>true</maven.test.failure.ignore>
		<log4j.version>1.2.17</log4j.version>
	</properties>

	<repositories>
		<repository>
			<snapshots>
				<enabled>false</enabled>
			</snapshots>
			<id>central</id>
			<name>bintray</name>
			<url>https://mvnrepository.com</url>
		</repository>
	</repositories>
	

	<dependencies>
		<dependency>
			<groupId>org.apache.commons</groupId>
			<artifactId>commons-lang3</artifactId>
			<version>3.6</version>
		</dependency>
		
		<dependency>
			<groupId>net.serenity-bdd</groupId>
			<artifactId>serenity-core</artifactId>
			<version>${serenity.version}</version>
		</dependency>
		<dependency>
			<groupId>net.serenity-bdd</groupId>
			<artifactId>serenity-cucumber</artifactId>
			<version>${serenity.cucumber.version}</version>
		</dependency>
		<dependency>
			<groupId>log4j</groupId>
			<artifactId>log4j</artifactId>
			<version>${log4j.version}</version>
		</dependency>
		<dependency>
			<groupId>org.slf4j</groupId>
			<artifactId>slf4j-simple</artifactId>
			<version>1.7.7</version>
		</dependency>
		<dependency>
			<groupId>net.serenity-bdd</groupId>
			<artifactId>serenity-junit</artifactId>
			<version>1.0.47</version>
		</dependency>
		<dependency>
			<groupId>junit</groupId>
			<artifactId>junit</artifactId>
			<version>4.12</version>
			<scope>test</scope>
		</dependency>
		<dependency>
			<groupId>org.assertj</groupId>
			<artifactId>assertj-core</artifactId>
			<version>1.7.0</version>
		</dependency>
	</dependencies>
	<build>
		<plugins>
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-surefire-plugin</artifactId>
				<version>2.18.1</version>
				<configuration>
					<systemPropertyVariables>
						<file>kpn.373-sit.apk</file>
						<appium.host>127.0.0.1</appium.host>
						<appium.port>4723</appium.port>
						<implicit.wait>200</implicit.wait>
						<platform.name>Android</platform.name>
						<device.name>4f5e98fe</device.name>
						<no.reset>true</no.reset>
						<appium.log>info</appium.log>
					</systemPropertyVariables>
					<skip>true</skip>
				</configuration>
			</plugin>
			<plugin>
				<artifactId>maven-failsafe-plugin</artifactId>
				<version>2.18.1</version>
				<configuration>
					
					<includes>
						<include>**/Run*.java</include>
						<include>**/test/**/*.java</include>
						<include>**/features/**/*.java</include>
					</includes>
					<systemProperties>
						<surefire.rerunFailingTestsCount>${surefire.rerunFailingTestsCount}</surefire.rerunFailingTestsCount>
						<surefire.rerunFailingTestsCount>${surefire.rerunFailingTestsCount}</surefire.rerunFailingTestsCount>
					</systemProperties>
					<encoding>${project.build.sourceEncoding}</encoding>
					<argLine>-Xmx512m</argLine>
				</configuration>
				<executions>
					<execution>
						<goals>
							<goal>integration-test</goal>
							<goal>verify</goal>
						</goals>
					</execution>
				</executions>
			</plugin>
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-compiler-plugin</artifactId>
				<version>3.2</version>
				<configuration>
					<fork>true</fork>
					<executable>C:\Program Files\Java\jdk1.8.0_144\bin\javac.exe</executable>
					
				</configuration>
			</plugin>
			<plugin>
				<groupId>net.serenity-bdd.maven.plugins</groupId>
				<artifactId>serenity-maven-plugin</artifactId>
				<version>${serenity.maven.version}</version>
				<executions>
					<execution>
						<id>serenity-reports</id>
						<phase>post-integration-test</phase>
						<goals>
							<goal>aggregate</goal>
						</goals>
					</execution>
				</executions>
			</plugin>
		</plugins>
	</build>
</project>
